### PR TITLE
BF Team Review

### PR DESCRIPTION
**Pull request by silochad**
# Review 1.
- [x] **fromToken / toToken vs. tokenIn / tokenOut**
  => ask team
- [x] **What to call `wellFunction()`?**
  - "curve, quoter, strategy"
  - priceFunction
  - Doesn't have to be a curve, could be an order
- [x] **RemoveLiquidityOneToken: tokenAmountOut -> amountOut**
  => "all the liquidity events and function signatures included token and lp"
  => symmetry elsewhere vs. symmetry with other RemoveLiquidity event
  => keep
- [x] **Swap amount names**
  getSwapIn  -> getAmountIn ?
  getSwapOut -> getAmountOut ?
  Pattern broken by getAddLiquidityOut. 
  => ask team: add `Amount` label?
  => ask team: is the asymmetry with `getSwapIn => swapFrom` a problem?
- [x] **addLiquidity: `minAmountOut -> minLpAmountOut` ?**
    => @silochad to refactor
- [x] **removeLiquidityOneToken: `token -> tokenOut`**
    => @silochad to refactor
- [x] **removeLiquidityOneToken: should `lpAmountIn` be the first parameter?**
    => @silochad to refactor
- [x] **`getRemoveLiquidityImbalanced -> getRemoveLiquidityImbalancedIn`**
    => ask team: should we add "amount" to all getters? (see above question)
    => definitely should add the `In` word for symmetry
- [x] **{IWell.initialize} doesn't exist**
    => @silochad to add a comment in Well.sol
- [x] **{IWell.well} doesn't exist**
    => @silochad to add function to IWell
- [x] **Why are vars/constructor at the bottom of ImmutableTokens?**
    => @silochad  to reorder items
- [ ] **`updatePumpsAndgetSwap`**
  - naming
  - uses iToken jToken
  - getSwapWithPump
  - getSwapAfterPump
  - updatePumpAndGetSwap
  - prepareSwap
  - "pump" = call "update" on pump, may get renamed
- [x] **`getIJ`**
  comments refers to "fromToken" "toToken"
    => @silochad  to refactor
- [x] **`getJ`**
  => fixed on latest master, merge
- [x]  do `i, j` need to be uint256 if # of tokens is limited? would it be more efficient to use a smaller uint type?
  => doesn't matter, uint8 takes a whole slot for memory/calldata
- [x]  Well.sol getLpTokenSupply() and getBalance(): wf vs _wellFunction
    use _wellFunction
    => @silochad  to refactor
- [x]  Well.sol well():
  Ordering should match the funcs below.
  `well()` should come after the functions so it's easier to decipher that we're using the overridden versions
  => @silochad to refactor
- [x]  constructor naming uses `_function` instead of `_welFunction`
    => keep it
- [x]  "Updates the Pump with the previous balances."
  Since this takes in arbitrary balances, should we remove the "previous"?
  The part that defines "previous" is the function which calls updatePump
  => @silochad to fix
- [x] "token" param inputs refer to addresses, but outputs are indexed based on tokens(). should we say ` @param token The token *address* to ...`

# Notes
- [x] natspec doc research: if a function implements an interface, does it automatically get the interface's comments, or do we need to explicitly use `inheritdoc`?
  getBalance vs getBalanceJ
    - getBalance always returns positive because it returns the new balance

  calculateSwap
    - positive amountIn = adding i, removing j
    - negative amountIn = adding j, removing i
    - before - after
    - amountOut is how much of j the user takes out
      - negative = user putting in
      - positive = user pulling out

### CODE REVIEW

**Comment by silochad** [comment]
# Review 2.

- [ ] Where are constraints on token construction set up? It looks right now like I could instantiate with `_tokens = [0xBEAN, 0x0000, 0xUSCD]`

**Comment by silochad** [comment]
# Review 3

- [ ] @Brean0: Why token addresses > indices?

**Comment by Brean0** [comment]
re: LibMath, using gaussian saves around 15k gas per liquidity in/out

**Comment by silochad** [comment]
# Review 4

- [FixedPointMathLib.sol](https://github.com/transmissions11/solmate/blob/main/src/utils/FixedPointMathLib.sol#L164)
- [Square roots by subtraction](https://studylib.net/doc/7921494/square-roots-by-subtraction---jarvis--frazer)
- [FlashBot.sol implementation](https://github.com/paco0x/amm-arbitrageur/blob/master/contracts/FlashBot.sol#L340)
- [Uniswap V3 Libraries](https://github.com/Uniswap/v3-core/tree/main/contracts/libraries)
  - [SqrtPriceMath](https://github.com/Uniswap/v3-core/blob/main/contracts/libraries/SqrtPriceMath.sol)
- [BancorFormula](https://github.com/bogatyy/bancor/blob/master/solidity/BancorFormula.sol#L10) uses some interesting table-based tactics

File: src/libraries/LibMath.sol [hunk 1]
```diff
@@ -19,13 +19,27 @@ library LibMath {
...
+        if(n % 2 == 0) {
+            if (n == 2) return sqrt(a); // shortcut for square root
+            if (n == 4) return sqrt(sqrt(a));
+            if (n == 6) return sqrt(sqrt(sqrt(a)));
+            if (n == 8) return sqrt(sqrt(sqrt(sqrt(a))));
+            if (n == 10) return sqrt(sqrt(sqrt(sqrt(sqrt(a)))));
+            if (n == 12) return sqrt(sqrt(sqrt(sqrt(sqrt(sqrt(a))))));
+            if (n == 14) return sqrt(sqrt(sqrt(sqrt(sqrt(sqrt(sqrt(a)))))));
+            if (n == 16) return sqrt(sqrt(sqrt(sqrt(sqrt(sqrt(sqrt(sqrt(a))))))));
+        }
```
Comment (Brean0) [inline src/libraries/LibMath.sol:42]: Need to write tests for n/2 cases

**Pull request merged**